commons-macros/src/main/scala/com/avsystem/commons/macros/serialization/GenCodecMacros.scala

```diff
@@ -258,30 +253,85 @@ class GenCodecMacros(ctx: blackbox.Context) extends CodecMacroCommons(ctx) with
       def generatedWrite(sym: Symbol) =
         q"writeField(${nameBySym(sym)}, output, ${mkParamLessCall(q"value", sym)}, ${depNames(sym)})"
 
+      def readField(param: ApplyParam) = {
+        val fieldName = nameBySym(param.sym)
+        val defaultValueTree = param.defaultValue.orElse(q"fieldMissing($fieldName)")
+        if (!isScalaJs) q"${optName(param)}.getOrElse($defaultValueTree)"
+        else {
+          if (param.valueType != weakTypeOf[Unit])
+            q"""
+              if (!_root_.scala.scalajs.js.isUndefined(elements.selectDynamic($fieldName))) {
+                elements.selectDynamic($fieldName).asInstanceOf[${param.valueType}]
+              } else ($defaultValueTree).asInstanceOf[${param.valueType}]
+            """
+          else q"()"
+        }
+      }
+
+      def generateReadObject =
+        if (!isScalaJs) {
+          q"""
+            def readObject(input: $SerializationPkg.ObjectInput, ..$oooFieldsParam): $tpe = {
+              ..${params.map(varDeclaration)}
+              while(input.hasNext) {
+                val fi = input.nextField()
+                fi.fieldName match {
+                  case ..${params.map(fieldReadCase)}
+                  case _ => fi.skip()
+                }
+              }
+              ${applier(params.map(p => p.asArgument(readField(p))))}
+            }
+          """
+        } else {
+          q"""
+            def readObject(input: $SerializationPkg.ObjectInput, ..$oooFieldsParam): $tpe = {
+              val elements: _root_.scala.scalajs.js.Dynamic = _root_.scala.scalajs.js.Object().asInstanceOf[_root_.scala.scalajs.js.Dynamic]
```

Comment:
I wonder how using a map or dictionary affects performance. I'll try to do some benchmarks in JVM - is there any blessed way of doing benchmarks for JS?
If the map/dictionary approach happens to be not slower for JVM then we could get rid of duplication and use it everywhere.

Reply:
Implementation with HashMap was about 2 times slower.

Reply:
I meant to compare the original variable based solution with map based solution.